### PR TITLE
Add ability to reload a file

### LINKED ISSUE #13212: Reset or refresh opened file
### Check for existing issues

- [X] Completed

### Describe the feature

(Sorry if dupe, searched pretty hard and couldn't find anything.)

Zed will notice when a file has updated out-of-band and warn me when I try to save, which is great. But how can I do the equivalent to Vim's `:e!` to reset to the current file's content?

Even better would be for a buffer which was initially unmodified (i.e. saved to disk) to just automatically refresh when it notices the file changed (i.e. due to a Git branch switch). It doesn't seem like that happens now. It seems like it prefers to retain the old content and deem it "modified" (state != disk) and then warns me about closing with unsaved changes. (I could be wrong here, sorry if so.) (**edit:** just saw this working, so not sure what gave me that impression)

Side note: I have the tab bar disabled locally for a more Vim-like experience, otherwise I end up with a million tabs duped between all my panels. So not seeing the file modification state might be a factor in my UX here.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by schpet**
Closes #13212

Release Notes:

- Added reload command
- vim: Added `:e[dit]`, `:e[dit]!` which calls reload

### CODE REVIEW

**Comment by ConradIrwin** [comment]
@schpet Thanks for this!

We have the code to reload already - for when you click "Discard" in the case of conflict on save (https://github.com/zed-industries/zed/blob/c83d007138587d832c5ce3a13c8cb99913c016be/crates/editor/src/items.rs#L788). 

This will ensure it also works for remote projects and should resolve the file status (though I'm less sure how well it deals with cursor position, as it's currently only called when an alert is open :D).

To make it work for vim, you can add it to the commands here: https://github.com/zed-industries/zed/blob/c83d007138587d832c5ce3a13c8cb99913c016be/crates/vim/src/command.rs#L434

File: crates/vim/src/command.rs [hunk 1]
```diff
@@ -604,6 +604,8 @@ fn generate_commands(_: &AppContext) -> Vec<VimCommand> {
         VimCommand::new(("$", ""), EndOfDocument),
         VimCommand::new(("%", ""), EndOfDocument),
         VimCommand::new(("0", ""), StartOfDocument),
+        // TODO: vim prints "E37: No write since last change (add ! to override)" if you have unsaved changes and you run :edit without the bang
+        VimCommand::new(("e", "dit"), editor::actions::ReloadFile).bang(editor::actions::ReloadFile),
```
Comment (schpet) [inline crates/vim/src/command.rs:608]: looking for suggestions on how to handle this. in vim, `:edit` will not throw away your changes, but `:edit!` will. to quote https://vimhelp.org/editing.txt.html#:edit

```
:e[dit] [++opt] [+cmd]	Edit the current file.  This is useful to re-edit the
			current file, when it has been changed outside of Vim.
			This fails when changes have been made to the current
			buffer and 'autowriteall' isn't set or the file can't
			be written.
			Also see ++opt and +cmd.

							:edit! discard
:e[dit]! [++opt] [+cmd]
			Edit the current file always.  Discard any changes to
			the current buffer.  This is useful if you want to
			start all over again.
			Also see ++opt and +cmd.

							:edit_f
```

i don't think i ever run :e without any arguments in vim, always `:e!`, is there a way to only map `:e!` `:edit!` in zed? or will that require changes to VimCommand ?

File: crates/editor/src/editor.rs [hunk 2]
```diff
@@ -6153,6 +6153,34 @@ impl Editor {
...
+                        if let Some(transaction) = transaction {
+                            if !buffer.is_singleton() {
+                                buffer.push_transaction(&transaction.0, cx);
+                            }
+                        }
+                    })
+                    .ok();
+            });
+        }
+    }
```
Comment (schpet) [inline crates/editor/src/editor.rs:6182]: i'm going to attempt to figure this out but if someone happens to look at this and knows what i need to do to call the [existing reload function](https://github.com/schpet/zed/blob/b5215fc7eceb3e27b9e09221d533dba0a553981c/crates/editor/src/items.rs#L788) i'm interested in learning (i don't know rust very well at all)

File: crates/vim/src/command.rs (hunk 1 above)
Comment (ConradIrwin) [reply crates/vim/src/command.rs:608]: We can either keep them the same (as you have done), and improve it later; or we could add another action that is just a no-op. I think it'd be better to keep them the same (until we add argument support and we get another try).

File: crates/editor/src/editor.rs (hunk 2 above)
Comment (ConradIrwin) [reply crates/editor/src/editor.rs:6182]: You should just be able to call `self.reload()` in your action callback. Does that not work?

Comment (schpet) [reply crates/editor/src/editor.rs:6182]: oh! thanks. https://github.com/zed-industries/zed/pull/18395/commits/c3a72e18b3ba1c4ba4b21782432d1485e90a86a4 attempts to do that. i am curious if

- is `Item as WorkspaceItem` legit? 
- is `let _ = self.reload(project, cx);` legit or should i try to do something with the result

**Comment by schpet** [comment]
@cla-bot check

File: crates/editor/src/editor.rs (hunk 2 above)
Comment (schpet) [reply crates/editor/src/editor.rs:6182]: oh: also wondering if project.clone() makes sense

Comment (ConradIrwin) [reply crates/editor/src/editor.rs:6182]: You probably need the clone to keep the borrow checker happy (`=` may not do what you might expect in rust https://cirw.in/blog/rust-assignment).

With the result you should be able to call `.notify_err()` on the result to have it show up:

```rust
 pub fn reload_file(&mut self, _: &ReloadFile, cx: &mut ViewContext<Self>) {
        let Some(workspace) = self.workspace() else {
           return;
        }
        let Some(project) = self.project.clone() else {
            return;
        }
        self.reload(project, cx).notify_err(workspace, cx)
    }
```

(though I haven't tried running it).

Comment (schpet) [reply crates/editor/src/editor.rs:6183]: huh! really appreciate the link and the post. 

landed on 1b1a690 - `self.reload(project, cx).detach_and_notify_err(cx);`

**Comment by alexanderjeurissen** [comment]
@schpet Great work on the PR, what is blocking us from shipping this ? how can we move this along ?

**Comment by schpet** [comment]
@alexanderjeurissen thanks! i'm waiting for the original reviewer to get back from a trip before following up with them. otherwise this just needs a set of discerning eyes and access to the merge button.

**Comment by ConradIrwin** [comment]
Thanks for this, and sorry for not hitting merge before disappearing for two weeks!

**Pull request merged**